#40 refactor(build-system): reduce gruntfile verbosity, drop grunt-if, allow custom build

1138-4EB opened the pull request:
# Reduce verbosity in `gruntfile.js`

`build*Binary` tasks are all equal, except `BUILD_GOOS`, `BUILD_GOARCH` and the executable name. I moved the instructions to a separate bash file (`build/build_cross-platform.sh`) which takes two args. Then, I replaced all the tasks with a single function-based one: https://stackoverflow.com/questions/20799069/grunt-grunt-shell-and-command-argument

Furthermore, the existence of `dist/portainer(.exe)` can be checked with `grunt.file.isFile`: https://nodejs.org/docs/latest/api/fs.html#fs_class_fs_stats . As a result, the dependency on `grunt-if`` can be dropped:

``` javascript
buildBinary: { command: function (p, a) {
  var file = 'dist/portainer'; 
  if ( p === "windows" ) { file = 'dist/portainer.exe'; }   
  if (grunt.file.isFile(file)) { return 'echo \'BinaryExists\'';
  } else {                      return 'build/build_cross-platform.sh ' + p + ' ' + a; }
}},  
```
Then, `if:<platform><Arch>BinaryNotExist` tasks are replaced with `'shell:buildBinary:<platform>:<Arch>`.

The very same approach can be applied to `release-<platform>-<arch>` tasks, which are replaced with a single one named `release`:

``` javascript
  grunt.task.registerTask('release', 'release:platform:arch', function(p, a) {
	var cp = 'copy';
	if ( p === "linux" && ( a === "386" || a === "amd64" ) ) { cp = 'copy:assets'; }
	grunt.task.run(['config:prod', 'clean:all', 'shell:buildBinary:'+p+':'+a, 'before-copy', cp, 'after-copy' ]);
  });
```
https://gruntjs.com/api/grunt.task
https://gruntjs.com/configuring-tasks

where

``` javascript
  grunt.registerTask('before-copy', [
    'html2js',
    'useminPrepare:release',
    'recess:build',
    'concat',
    'clean:tmpl',
    'cssmin',
    'replace',
    'uglify', 
  ]);
  grunt.registerTask('after-copy', [
    'filerev',
    'usemin',
    'clean:tmp' 
  ]);
```
Note these are also used in `build` and `build-webapp` to further reduce verbosity.

# Reduce verbosity in `build.sh`

Despite `PLATFORM` and `ARCH` being defined, these are always used as `PLATFORM-ARCH`. Therefore couples (tags) can explicitly defined in that form. On top of that, an array can hold multiple tags:

```
platforms='linux-amd64 linux-386 linux-arm linux-arm64 linux-ppc64le darwin-amd64 windows-amd64'
 
for tag in ${platforms}; do
  grunt release:`echo "$tag" | tr '-' ':'`
  if [ `echo $tag | cut -d \- -f 1` == "linux" ]; then build_and_push_images "$tag"; fi
  build_archive "$tag"
done
```
- `build_and_push_images` and `build_archive` are modified to accept a single argument.
- Due to previous modifications, calls to grunt must be `release:<platform>:<arch>`. Then, `echo` and `tr` are used to replace `-` with `:`. 

Furthermore, during development, `build.sh` is modified in order to allow custom build commands. If the first four characters in "<VERSION>" are "echo", "$@" is passed to "bash -c":

``` bash
Usage: build.sh <VERSION>
       build.sh "echo 'Custom' && <BASH COMMANDS>"
```

Close #40

deviantony commented:
Brilliant ! I'll have a look as soon as I can.

deviantony commented on build.sh:
```diff
@@ -1,79 +1,47 @@
-grunt release-${PLATFORM}-${ARCH}
-build_archive ${PLATFORM} ${ARCH}
+function build_all() {
+  mkdir -pv /tmp/portainer-builds
```
Shouldn't that be replace with `mkdir -pv ${ARCHIVE_BUILD_FOLDER}` ?

1138-4EB replied:
It should. Changed it.

deviantony commented on build/build_cross-platform.sh:
```diff
@@ -0,0 +1,9 @@
+
+mkdir -p dist
+mv "api/cmd/portainer/$binary" dist/
+shasum "dist/$binary" > portainer-checksum.txt
```
I'm wondering if we should remove the `shasum` step as it's not used at the moment (plus the binary is platform dependent, see the section about CentOS in http://portainer.readthedocs.io/en/stable/contribute.html).

1138-4EB replied:
Indeed, in alpine I use `sha256sum`. It is a little trickier because shasum exists, so a ln is not possible.
I think that `sha256sum` is portable enough. However, since it is not used, I commented it. I didn't remove it, because it is a single line and it may be useful in the near future.

deviantony commented on gruntfile.js:
```diff
@@ -9,167 +9,59 @@ module.exports = function (grunt) {
-    'clean:tmp'
+    'clean:tmp' 
   ]);
+  grunt.task.registerTask('release', 'release:<platform>:<arch>', function(p, a) {
```
I do prefer readability over minimization, could you replace the `p` and `a` vars with meaningful variable names? Could you do the same for the code introduced below?

1138-4EB replied:
Although not so explicit, you have the meaning there:

```
'release', 'release:<platform>:<arch>', function(p, a)
```

The function is named 'release' and takes two arguments ('platfom' and 'arch'). Then, var 'p' is arg 'platform', and var 'a' is arg 'arch'.

However, if you still prefer it, I can replace every 'p' with 'platform and every 'a' with 'arch'.

deviantony commented on gruntfile.js:
```diff
@@ -9,167 +9,59 @@ module.exports = function (grunt) {
-  grunt.registerTask('run', ['if:linuxAmd64BinaryNotExist', 'build', 'shell:buildImage', 'shell:run']);
-  grunt.registerTask('run-dev', ['if:linuxAmd64BinaryNotExist', 'shell:buildImage', 'shell:run', 'watch:build']);
+  grunt.task.registerTask('watchBuild', 'watchBuild:<arg>', function(a) {
+	var b = a === '' ? 'run' : a;
```
See my comment above on readability. At first look, I can't understand what's this line is supposed to do.

1138-4EB replied:
Following the explanation above, 'a' is 'arg'.
'b' is just an auxliary var. Shall I name it `aux`?
Alternatively, I think that 'b' can be dropped:

```
  grunt.task.registerTask('watchBuild', 'watchBuild:<arg>', function(a) {
	grunt.task.run(['build', 'shell:buildImage', 'shell:'+ (a === '' ? 'run' : a) , 'shell:cleanImages']);
  });
```

deviantony commented on gruntfile.js:
```diff
@@ -388,7 +280,7 @@ module.exports = function (grunt) {
       build: {
         files: ['<%= src.js %>', '<%= src.css %>', '<%= src.tpl %>', '<%= src.html %>'],
-        tasks: ['build', 'shell:buildImage', 'shell:run', 'shell:cleanImages']
+        tasks: ['watchBuild:']
```
IMO we should only be able to trigger a build with watch for the local endpoint. No need for the SSL or Swarm setups at the moment since we can still add them as endpoints in the UI.

You can remove anything related to runSSL, runSwarm, buildSSL, etc...

1138-4EB replied:
I have not used Swarm and I was not confident enough to make any functional modification. According to you comment not, I removed lines 291-298. However, when doing so the task 'watchBuild' at line 59 (see above) is not required anymore. Shall I step back, remove that task, and leave a single 'watch:build:' task at line 283?

1138-4EB replied:
BTW, I don't understand the long comment at lines 284-289. Did you write it?

1138-4EB commented:
This last commit includes all the requested changes, except renaming 'p' to 'platform' and 'a' to 'arch'. Just let me know if you still think it is required.

BTW, note that deploying docker images to the hub is still included in `build.sh`, although we have talked in #938 about moving it to a new `deploy.sh`. If you agree, I'll leave this untouched and we will work on it in the other PR.

deviantony commented on gruntfile.js:
```diff
@@ -9,167 +9,55 @@ module.exports = function (grunt) {
+    'clean:tmp' 
   ]);
+  grunt.task.registerTask('release', 'release:<platform>:<arch>', function(p, a) {
+	var cp = ( p === 'linux' && ( a === '386' || a === 'amd64' ) ) ? 'copy:assets' : 'copy';
```
Why do we trigger the `copy:assets` task only for Linux builds?

1138-4EB replied:
No idea at all. Note that it is not only for Linux, because for 'linux-arm' `copy` is used.
Anyway, that's not something I changed. You can check the develop branch: https://github.com/portainer/portainer/blob/develop/gruntfile.js Lines 30, 46, 63, 80, 97, 114, 131, 148 and 165.

deviantony replied:
Ah, let met investigate that.

deviantony replied:
Ok so after investigation, the `copy` task is called within `build` and `build-webapp` only.

`build` is used with `grunt run-dev` and `build-webapp` is only used for CI with codefresh.

1138-4EB replied:
I got lost with the reference to `copy:bundle`, because I don't see when is it called.

> Every release task should only use the copy:assets task.

I replaced `var cp = ( p === 'linux' && ( a === '386' || a === 'amd64' ) ) ? 'copy:assets' : 'copy';` with `copy:assets` for every release.

> Note: as an enhancement minified vendor and angular libraries should only be used during release, unminified libraries should be used during development.

That's the actual behaviour, isn't it?

deviantony commented:
Side note, #972 was open recently and will probably be merged before this one.

1138-4EB commented:
> Side note, #972 was open recently and will probably be merged before this one.

Most of the changes do not conflict at all. The only conflicting one is replacing `recess:build` and `cssmin` with `postcss:build`. Shall I merge `G07cha:replace-recess-with-postcss` with `1138-4EB:refactor40-grunt-tasks`, so that changes corresponging to both PRs are shown here?

deviantony commented:
> Most of the changes do not conflict at all. The only conflicting one is replacing recess:build and cssmin with postcss:build. Shall I merge G07cha:replace-recess-with-postcss with 1138-4EB:refactor40-grunt-tasks, so that changes corresponging to both PRs are shown here?

Not yet please, I'll test and merge #972 first

deviantony commented:
@1138-4EB #975 has been merged, could you rebase?

1138-4EB commented:
@deviantony , rebased. And I added the last change you commented above: to always use `copy:assets` in releases, instead of `copy` in non-linux builds. I don't know what to do with `copy:bundle` tho.

1138-4EB commented:
> I'm pretty sure that the following task can be removed too:
> `grunt.registerTask('run', ['shell:buildBinary:linux:amd64', 'build', 'shell:buildImage', 'shell:run']);`
> It seems not to be used.

Removed. And renamed run-dev to run.

> Sorry, missed that, which comment are you talking about?

```
    watch: {
      all: {
        files: ['<%= src.js %>', '<%= src.css %>', '<%= src.tpl %>', '<%= src.html %>'],
        tasks: ['default', 'timestamp']
      },
      build: {
        files: ['<%= src.js %>', '<%= src.css %>', '<%= src.tpl %>', '<%= src.html %>'],
        tasks: ['build', 'shell:buildImage', 'shell:run', 'shell:cleanImages']
        /*
        * Why don't we just use a host volume
        * http.FileServer uses sendFile which virtualbox hates
        * Tried using a host volume with -v, copying files with `docker cp`, restating container, none worked
        * Rebuilding image on each change was only method that worked, takes ~4s per change to update
        */		
      }
    }
```

>Actually, if you look at the src.jsVendor and src.cssVendor vars (as well as the concat.angular task), you'll see that they are always using minified versions when they're available. I was thinking that maybe we should use non-minified libraries during development.

I'm having a look at it.

> Perfect, copy:bundle will be triggered as a part of the copy task in build, that is the only usage of this task.

Oh, I think I get it now: when you run `copy` without an argument/option, both `copy:bundle` and `copy:assets` are executed.

deviantony commented:
> Removed. And renamed run-dev to run.

I'd like to keep the task named `run-dev` to avoid breaking other contributor's dev flow.

> Oh, I think I get it now: when you run copy without an argument/option, both copy:bundle and copy:assets are executed.

Yes, when you don't specify a subtask, all subtasks are run.

> /*
        * Why don't we just use a host volume
        * http.FileServer uses sendFile which virtualbox hates
        * Tried using a host volume with -v, copying files with `docker cp`, restating container, none worked
        * Rebuilding image on each change was only method that worked, takes ~4s per change to update
        */	

This was part of the build system of a previous project. It's pretty old, and I've always thought that it would be a better idea for local development.

At the moment, the watch-build flow is the following (triggered after each file saved):

* Build the app (generate JS/CSS/fonts/etc in /dist)
* Build a Docker image 
* Run the previously built Docker image
* Clean old containers/images

It'd be interesting to see if we could somehow just build the app after save and mount the app files inside a Docker container instead of re-creating an image & container each time. See what I mean?

1138-4EB commented:
> I'd like to keep the task named run-dev to avoid breaking other contributor's dev flow.

OK, I left `run-dev`.

> It'd be interesting to see if we could somehow just build the app after save and mount the app files inside a Docker container instead of re-creating an image & container each time. See what I mean?

I think I do. That 'Build the app' step includes building the go part or just `build-webapp`?

---

Note that this last commit converts several `dest: '<dest>', src: '<src>',` declarations to the more compact form `<dest>: <src>`.

On top of that, it includes a first approach to not use minified versions in dev. I just defined an envvar, which is either set to `.min` or left empty.

deviantony commented:
> I think I do. That 'Build the app' step includes building the go part or just build-webapp?

Both, if the binary is not present it needs to be built.

> Note that this last commit converts several dest: '<dest>', src: '<src>', declarations to the more compact form <dest>: <src>.

I'm good with that.

> On top of that, it includes a first approach to not use minified versions in dev. I just defined an envvar, which is either set to .min or left empty.

Interesting, is it working yet? Can I test it?

1138-4EB commented:
Not yet. I'm checking it right now.

deviantony commented:
Good, note that you can also ping me on Slack if you have questions and you want this discussion to be more reactive.

1138-4EB commented:
I think it is ready now. I undid the `src:` and `dest:` modification, because some plugins do not properly handle it. I'll include it in #938, which let's me test changes remotely. I'll also investigate there about using a docker container for 'hot' builds.

About this PR, I think that the only renaiming issue is that some of the files to which the `.min` suffix is conditionally applied might not exist. And I think that `concat` does not throw errors if some of the files are not present. Can we know for sure that there is a `file.ext` in the same path as `file.min.ext`? For example, is `bower_components/moment/min/moment.js` present?

deviantony commented:
> About this PR, I think that the only remaining issue is that some of the files to which the .min suffix is conditionally applied might not exist. And I think that concat does not throw errors if some of the files are not present. Can we know for sure that there is a file.ext in the same path as file.min.ext? For example, is bower_components/moment/min/moment.js present?

You're right, some libraries might not provide minified files, we need to handle this case.

deviantony commented:
Just gave a try to the latest version:

```
$ grunt run-dev
Running "shell:buildBinary:linux:amd64" (shell) task
/bin/sh: 1: build/build_in_container.sh: not found
Warning: Command failed: build/build_in_container.sh linux amd64
/bin/sh: 1: build/build_in_container.sh: not found
 Use --force to continue.

Aborted due to warnings.
```

Something is missing :)

1138-4EB commented:
Should be fixed now. It was a C&P error on my side.

deviantony commented:
@1138-4EB could you rebase? We need to tackle this one before #994 and #938

1138-4EB commented:
There were two naming related issue:

- `shell:buildImage` is not used in `grunt run-dev` any more after #997. However, the issue remains, because the container expects the binary to be named `portainer`.
  - We can either move/rename `portainer-linux-amd64` or change the argument to he container. I chose to change the argument.
- In this PR `buildBinary` does not produce any binary named `portainer`. Therefore, it checks for the existence of `dist/portainer-linux-amd64`. However, `clean:app` removes all the content of `dist` but `dist/portainer`. I changed clean:app so that `dist/portainer*` are preserved.

1138-4EB commented:
`shell:buildImage` and `shell:cleanImages` where only used in `run-dev`. Then, I removed them from the gruntfile. On top of that, since images are built in `build.sh`, I added que command corresponding to `cleanImages` there.

Side note: `grunt run-dev` fails if a container named 'portainer' is not already running. I don't think this is a problem. But it might be convenient to document this somewhere.

deviantony commented on gruntfile.js:
```diff
@@ -375,135 +241,35 @@ module.exports = function (grunt) {
+                 }
+      },  
       run: {
         command: [
```
You can replace

```
        command: [
          'docker stop portainer',
          'docker rm portainer',
          'docker run -d -p 9000:9000 -v $(pwd)/dist:/app -v /tmp/portainer:/data -v /var/run/docker.sock:/var/run/docker.sock:z --name portainer centurylink/ca-certs /app/portainer-linux-amd64 --no-analytics -a /app'
        ].join(';')
```

With:

```
        command: [
          'docker rm -f portainer',
          'docker run -d -p 9000:9000 -v $(pwd)/dist:/app -v /tmp/portainer:/data -v /var/run/docker.sock:/var/run/docker.sock:z --name portainer centurylink/ca-certs /app/portainer-linux-amd64 --no-analytics -a /app'
        ].join(';')
```

deviantony commented on gruntfile.js:
```diff
@@ -375,135 +241,35 @@ module.exports = function (grunt) {
-      buildSsl: {
-        files: ['<%= src.js %>', '<%= src.css %>', '<%= src.tpl %>', '<%= src.html %>'],
-        tasks: ['build', 'shell:buildImage', 'shell:runSsl', 'shell:cleanImages']
+        /*
```
This whole block of comment can now be removed (solved via #997)

1138-4EB commented:
> Please have a look at my comments to remove the useless comments and update the shell:run task.

Done.

> I had a look at the whole gruntfile (which definitely looks better :-)) and I think that the watch:all task is never used and can be removed.

Removed `watch:all` and, consequently, `timestamp`.

> Also, I think that the build task should always be executed when using grunt run-dev, at the moment if I execute grunt run-dev I need to trigger a file change to trigger the build.

I now replaced `'shell:buildBinary:linux:amd64'` with `'build'`.

PS: I rebased on top of develop.

deviantony commented:
Thanks @1138-4EB :+1: 

Now, to the next step ;)
